### ISSUE TITLE
TaskLocal not always restoring context after execution

### ISSUE THREAD

**Issue by oleg-py** [monix/monix#624]
I finally managed to reproduce the issue I was talking about on gitter.

The app below crashes with `MatchError` because variable becomes `true` on 2nd iteration (on my machine).

Happens on both RC1 and #613 (commit # 63044eb)

```scala
import monix.execution.Scheduler.Implicits.global
import monix.eval._
import scala.concurrent.duration._

object Fakes extends App {
  implicit val opts = Task.Options(false, true)
  val local = TaskLocal[Boolean](false).runSyncUnsafeOpt(Duration.Inf)

  def attempt = local.read flatMap {
    case false => local.bind(true) {
      Task.unit
    }
  }

  for (_ <- 1 to 1000) {
      attempt.runSyncUnsafeOpt(Duration.Inf)
  }
}
```
Ironically, this works when LCP is *disabled*

/cc @leandrob13

**Comment by leandrob13** [comment]
@oleg-py got it, will test and debug asap

**Comment by alexandru** [comment]
So this sample should probably work because it's on the same thread, don't know what is wrong here; but it got me thinking ... your usage of `runSyncUnsafeOpt` in that loop is incompatible with the chosen model and this can't work for all tasks.

The problem gets more complicated when you introduce forks:

```scala
  def attempt = local.read flatMap {
    case false => local.bind(true) {
      Task.shift // <-- fork
    }
  }
```

In such a case the restore will happen on another thread entirely, captured in a `ThreadLocal`, which will never be seen from `main`. Basically the restore happens for the bind continuation of that `Task`, so the loop that should work is this:

```scala
def loop(n: Int = 1000): Task[Unit] =
  if (n > 0) attempt.flatMap(_ => loop(n - 1))
  else Task.unit
```

Is this a use case we care about?

If yes, then we cannot deliver on `TaskLocal`. If no, then we need to warn people in the docs, but if somebody as capable as you @oleg-py couldn't see the problem in this, then I'm afraid that `TaskLocal` is too error prone.

**Comment by alexandru** [comment]
Btw, our implementation is inspired by the implementation of Twitter's `Future` and `Local`.

But it has one big difference. Back when we implemented this, @leandrob13 used an immutable `Map` in that `ThreadLocal` which I liked very much, however Twitter's implementation uses a mutable `Array`.

I'm now thinking that we might have made the wrong decision and that we also need a mutable `Array`.

**Comment by oleg-py** [comment]
@alexandru I used `shift` initially in my testing, but as I went on simplifying, `unit` was enough.

Honestly, my understanding of concurrency issues is somewhat superficial. I prefer to rely on good practices (immutables, pure functions) and abstractions (Task) to avoid having them 😅.

As I mentioned on gitter, I encounter similar problem when managing connections in oleg-py/quill-monix. The `Task` I produce is eventually run by http4s server, so I have little control over how it gets evaluated.

`TaskLocal` is a key for supporting transactions in `quill-monix` which has the potential to be a decent alternative for doobie in FP setting. It also has other uses like per-request logging data (and horrible ones, like avoiding passing arguments to functions :).

**Comment by alexandru** [comment]
Yeah. I'm a little busy at work, but will play around with it this week. Maybe we need to go back to the basics. I'm curious how it behaves with a mutable Array in it, like Twitter is doing.

I somehow thought we outsmarted them, but I think they outsmarted us 😀

**Comment by leandrob13** [comment]
Getting back from holidays. Should I pause this until @alexandru tests the mutable Array alternative?

**Comment by leandrob13** [comment]
@oleg-py I tested and it seems that the problem lies within the `bindL` implementation based ob `bracket`.

Reverted back to this:

```scala
def bind[R](value: A)(task: Task[R]): Task[R] =
    Task.suspend {
      val saved = ref.value
      ref.update(value)
      task.doOnFinish(_ => restore(saved))
    }
```

and it works as you expect. The problem is that `bracket` forces an async boundary by using the `onCancelRaiseError`:

```scala
def apply[A, B](
    acquire: Task[A],
    use: A => Task[B],
    release: (A, Either[Option[Throwable], B]) => Task[Unit]): Task[B] = {

    acquire.flatMap { a =>
      val next = try use(a) catch { case NonFatal(e) => Task.raiseError(e) }
      next.onCancelRaiseError(isCancel).flatMap(new ReleaseFrame(a, release))
    }
  }
```

Since the boundary is forced, the release is done in another thread instead of the current one. That is something that is ok with `bracket`.  Tested for #601 and the test I based on your reports passes. I can do the revert in #613 if you guys think it is ok.

**Comment by alexandru** [comment]
The async boundary in `onCancelRaiseError` is a light one, being managed by `TrampolineRunnable` instances pushed in the `Scheduler`, which knows how to execute them on the current thread.

That piece of logic never forks the execution.

What happens actually is that in the `TaskRunLoop` we are using `Local.bind` on executing stuff, restoring the previous context on the current thread after execution has completed or forked on another thread.

So on one of those `Async` values we end up doing:

```scala
// Saving the current context
savedLocals = Local.getContext()

// ...

Local.bind(savedLocals) {
  // Executing the bind continuation ...
  startFull(Now(value), context, callback, this, bFirst, bRest, runLoopIndex())
} // <-- restore happens afterwards
```

**Comment by alexandru** [comment]
This is also consistent with how the `TracingScheduler` works:

```scala
final class TracingRunnable(r: Runnable, context: Context = Local.getContext())
  extends Runnable {

  def run(): Unit = {
    val prev: Context = Local.getContext() // <-- saves current context
    Local.setContext(context) // <-- sets the context we had on scheduler.execute()
    try r.run() finally Local.setContext(prev) // <-- restores saved context
  }
}
```

**Comment by leandrob13** [comment]
@alexandru @oleg-py As we discussed in gitter, I came up with this that seems to fix the problem reported:

```scala
def bindL[R](value: Task[A])(task: Task[R]): Task[R] =
    Task.eval(ref.value).flatMap { saved =>
      value.bracket(v => ref.bind(v)(task))(_ => restore(saved))
    }
```

If this is approved I can add it in #613.

**Comment by leandrob13** [comment]
@oleg-py Nevermind, this breaks the other tests. I had a premature conclusion on this one.

**Comment by leandrob13** [comment]
@oleg-py I got it working for your example that uses `Task.unit` with `bind` and `bindL` keeping the bracket implementation:

```scala
def bindL[R](value: Task[A])(task: Task[R]): Task[R] =
    local.flatMap { r =>
      val saved = r.value
      value.bracket { v =>
        r.update(v)
        task.doOnFinish(_ => restore(saved))
      }(_ => restore(saved))
    }
```

If I changed your example above from `Task.unit` to `Task.shift` and it is still failing because of the reasons @alexandru gave above ( I tried to explain it (here)[https://gist.github.com/leandrob13/382df8a7b8cec321bf5d2791df240889] ). 

I don't think we can guarantee the cleanup of the original thread the Task starts from with our current status. This effort might require to keep an additional state of the context before and after a bind so the former can be restored on the original thread upon reaching an async boundary. It would mean to change the `LocalAwareCallback` I proposed. We can see if @oleg-py can do it in #635
